File: connectx-frontend/src/api/conversationApi.ts
import { apiRequest } from './apiClient';
import { Conversation } from '../types';

let inFlightConversations: Promise<Conversation[]> | null = null;

export const conversationApi = {
  getConversations: (): Promise<Conversation[]> => {
    if (inFlightConversations) return inFlightConversations;

    inFlightConversations = apiRequest<Conversation[]>('/conversations')
      .finally(() => {
        inFlightConversations = null;
      });

    return inFlightConversations;
  },

  createDirectConversation: (userId: number) =>
    apiRequest<Conversation>('/conversations/direct', {
      method: 'POST',
      body: JSON.stringify({ userId }),
    }),

  getConversationById: (conversationId: number) =>
    apiRequest<Conversation>(`/conversations/${conversationId}`),

  deleteConversation: (conversationId: number) =>
    apiRequest<string>(`/conversations/${conversationId}`, {
      method: 'DELETE',
    }),

  clearConversation: (conversationId: number) =>
    apiRequest<string>(`/conversations/${conversationId}/clear`, {
      method: 'POST',
    }),

  pinConversation: (conversationId: number) =>
    apiRequest<Conversation>(`/conversations/${conversationId}/pin`, {
      method: 'POST',
    }),

  unpinConversation: (conversationId: number) =>
    apiRequest<Conversation>(`/conversations/${conversationId}/unpin`, {
      method: 'POST',
    }),

  muteConversation: (conversationId: number, mutedUntil?: string) =>
    apiRequest<Conversation>(`/conversations/${conversationId}/mute`, {
      method: 'POST',
      body: JSON.stringify({ mutedUntil }),
    }),

  unmuteConversation: (conversationId: number) =>
    apiRequest<Conversation>(`/conversations/${conversationId}/unmute`, {
      method: 'POST',
    }),
};

File: connectx-android/app/src/main/java/com/connectx/app/data/remote/conversation/ConversationApi.kt
package com.connectx.app.data.remote.conversation

import com.connectx.app.core.network.model.ApiResponse
import com.connectx.app.data.remote.conversation.model.ConversationDto
import com.connectx.app.data.remote.conversation.model.CreateDirectConversationDto
import com.connectx.app.data.remote.conversation.model.MuteConversationRequest
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path

/**
 * Full Android-side contract for the backend's Conversation API, verified directly
 * against com.connectx.conversation.controller.ConversationController -- 13
 * endpoints, matching the N2.0 inspection's reported count exactly (no discrepancy
 * this time; N2.2's User undercount was the exception, not the rule).
 *
 * No pagination exists on this controller -- GET /conversations returns a plain
 * List<ConversationDto>, not a Page/Slice/cursor structure.
 *
 * Contract only -- no call site exists anywhere yet. All 13 endpoints require
 * authentication on the backend; no Authorization header handling is added here
 * (deferred to N3). No MessageApi/message models were introduced -- ConversationDto's
 * last-message fields are flat scalars, not a nested message object (see
 * ConversationDto.kt's doc comment).
 */
interface ConversationApi {

    @GET("api/v1/conversations")
    suspend fun getConversations(): ApiResponse<List<ConversationDto>>

    @POST("api/v1/conversations/direct")
    suspend fun createOrGetDirectConversation(
        @Body request: CreateDirectConversationDto
    ): ApiResponse<ConversationDto>

    @GET("api/v1/conversations/{conversationId}")
    suspend fun getConversationById(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>

    @DELETE("api/v1/conversations/{conversationId}")
    suspend fun deleteConversationForUser(@Path("conversationId") conversationId: Long): ApiResponse<String>

    @POST("api/v1/conversations/{conversationId}/clear")
    suspend fun clearConversationForUser(@Path("conversationId") conversationId: Long): ApiResponse<String>

    @POST("api/v1/conversations/{conversationId}/pin")
    suspend fun pinConversation(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>

    @POST("api/v1/conversations/{conversationId}/unpin")
    suspend fun unpinConversation(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>

    /** Backend body is optional (@RequestBody(required = false)) -- pass null for "mute indefinitely". */
    @POST("api/v1/conversations/{conversationId}/mute")
    suspend fun muteConversation(
        @Path("conversationId") conversationId: Long,
        @Body request: MuteConversationRequest?
    ): ApiResponse<ConversationDto>

    @POST("api/v1/conversations/{conversationId}/unmute")
    suspend fun unmuteConversation(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>

    @POST("api/v1/conversations/{conversationId}/archive")
    suspend fun archiveConversation(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>

    @POST("api/v1/conversations/{conversationId}/unarchive")
    suspend fun unarchiveConversation(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>

    @POST("api/v1/conversations/{conversationId}/mark-unread")
    suspend fun markConversationUnread(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>

    @POST("api/v1/conversations/{conversationId}/mark-read")
    suspend fun markConversationRead(@Path("conversationId") conversationId: Long): ApiResponse<ConversationDto>
}
